components/pd_client/src/util.rs

```diff
@@ -456,42 +581,90 @@ impl PdConnector {
+        let mut has_network_error = false;
+        let client_urls = peer.get_client_urls();
+        for ep in client_urls {
+            match self.connect(ep.as_str()).await {
+                Ok((client, address, forwarded_host, _)) => {
+                    info!("connected to PD member"; "endpoints" => ep);
+                    return Ok((Some((client, address, forwarded_host)), false));
+                }
+                Err(Error::Grpc(RpcFailure(RpcStatus { status, details: _ }))) => {
+                    if status == RpcStatusCode::UNAVAILABLE
+                        || status == RpcStatusCode::DEADLINE_EXCEEDED
+                    {
+                        error!("failed to connect to PD member"; "endpoints" => ep, "status_code" => ?status);
+                        network_fail_num += 1;
+                    }
+                }
+                Err(e) => {
+                    error!("failed to connect to PD member"; "endpoints" => ep, "error" => ?e);
+                }
             }
         }
-        None
+        let url_num = client_urls.len();
+        if url_num != 0 && url_num == network_fail_num {
+            has_network_error = true;
+        }
+        Ok((None, has_network_error))
     }
 
     pub async fn reconnect_leader(
         &self,
-        previous: &GetMembersResponse,
-        force: bool,
-    ) -> Result<Option<StubPair>> {
-        let resp = self.load_members(&previous).await;
+        leader: &Member,
+    ) -> Result<(Option<(PdClientStub, String, String)>, bool)> {
+        fail_point!("connect_leader", |_| Ok((None, true)));
+        let mut retry_times = MAX_RETRY_TIMES;
 
-        // Then try to connect the PD cluster leader.
-        match resp {
-            Some(resp) => {
-                if !force && resp == *previous {
-                    return Ok(None);
+        // Try to connect the PD cluster leader.
+        loop {
+            let (res, has_network_err) = self.connect_member(leader).await?;
+            match res {
+                Some((client, address, forwarded_host)) => {
+                    return Ok((Some((client, address, forwarded_host)), has_network_err))
                 }
-                match self.connect_member(resp.get_leader()).await {
-                    Some(client) => Ok(Some((client, resp))),
-                    None => Err(box_err!("failed to connect to {:?}", resp.get_leader())),
+                None => {
+                    if has_network_err && retry_times > 0 {
+                        let _ = GLOBAL_TIMER_HANDLE
+                            .delay(Instant::now() + Duration::from_secs(RETRY_INTERVAL_SEC))
+                            .compat()
+                            .await;
+                        retry_times -= 1;
+                        continue;
+                    }
+                    return Ok((None, has_network_err));
+                }
+            }
+        }
+    }
+
+    pub async fn reconnect_followers(
```

Comment:
I don't see why it needs to be a standalone function. Why not just keep it in `reconnect_leader`? Separating the two functions is inaccurate extraction and make code duplicated. There is no such connecting follower thing, what PD client does is always try to connect leader, either by connecting directly or forwarding by a follower. Connecting a follower is never its goal.